### PR TITLE
[docs] Merge LoRAs

### PR DESCRIPTION
**Pull request by stevhliu**
Continuation of #7110 to implement [option 3](https://github.com/huggingface/diffusers/pull/7110#pullrequestreview-1904781705) for combining all merge methods into one doc and cleaning up the other docs to focus on loading or basic usage.

### CODE REVIEW

File: docs/source/en/_toctree.yml [hunk 1]
```diff
@@ -18,7 +18,7 @@
   - local: tutorials/basic_training
     title: Train a diffusion model
   - local: tutorials/using_peft_for_inference
-    title: Inference with PEFT
+    title: Load LoRAs for inference
```
Comment (sayakpaul) [inline docs/source/en/_toctree.yml:21]: It makes a lot of sense now because we have made `peft` a required dependency to do all things LoRA in `diffusers` :)

File: docs/source/en/tutorials/using_peft_for_inference.md [hunk 2]
```diff
@@ -165,77 +147,6 @@ list_adapters_component_wise
 {"text_encoder": ["toy", "pixel"], "unet": ["toy", "pixel"], "text_encoder_2": ["toy", "pixel"]}
 ```
 
-## Compatibility with `torch.compile`
```
Comment (sayakpaul) [inline docs/source/en/tutorials/using_peft_for_inference.md:168]: Why is this going away? 😱

File: docs/source/en/tutorials/using_peft_for_inference.md [hunk 3]
```diff
@@ -165,77 +147,6 @@ list_adapters_component_wise
...
-## Saving a pipeline after fusing the adapters
-
-To properly save a pipeline after it's been loaded with the adapters, it should be serialized like so:
-
-```python
-pipe.fuse_lora(lora_scale=1.0)
-pipe.unload_lora_weights()
-pipe.save_pretrained("path-to-pipeline")
-```
+Diffusers also offers more efficient merging methods which you can learn about in the [Merge LoRAs](../using-diffusers/merge_loras) guide!
```
Comment (sayakpaul) [inline docs/source/en/tutorials/using_peft_for_inference.md:152]: I think this should be added as a "Tip" under the "Merge adapters" section. Also, the merging utilities come from `peft`. So, I would reword this a bit.

File: docs/source/en/using-diffusers/loading_adapters.md [hunk 4]
```diff
@@ -165,101 +165,14 @@ To unload the LoRA weights, use the [`~loaders.LoraLoaderMixin.unload_lora_weigh
 pipeline.unload_lora_weights()
 ```
 
-### Load multiple LoRAs
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/loading_adapters.md:168]: Do we have to include a note mentioning the guide on "merging adapters"?

File: docs/source/en/using-diffusers/merge_loras.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+prompt = "A bowl of ramen shaped like a cute kawaii bear, by Feng Zikai"
+image = pipeline(prompt, generator=generator, cross_attention_kwargs={"scale": 1.0}).images[0]
+image
+```
+
+<div class="flex justify-center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/diffusers/lora_merge_set_adapters.png"/>
+</div>
+
+## add_weighted_adapters
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/merge_loras.md:47]:
```suggestion
## add_weighted_adapter
```

File: docs/source/en/using-diffusers/merge_loras.md [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+
+> [!WARNING]
+> This is an experimental method that adds PEFTs [`~peft.LoraModel.add_weighted_adapter`] method to Diffusers to enable more efficient merging methods. Check out this [issue](https://github.com/huggingface/diffusers/issues/6892) if you're interested in learning more about the motivation and design behind this integration.
+
+The [`~peft.LoraModel.add_weighted_adapter`] method provides access to more efficient merging method such as [TIES and DARE](https://huggingface.co/docs/peft/developer_guides/model_merging). To use these merging methods, make sure you install PEFT from source and have the latest stable version of Diffusers installed.
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/merge_loras.md:52]: We don't have to install `peft` from the source anymore. The latest release has the utilities :)

File: docs/source/en/using-diffusers/merge_loras.md [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+
+```bash
+pip install git+https://github.com/huggingface/peft.git
+pip install -U diffusers
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/merge_loras.md:56]:
```suggestion
pip install -U diffusers peft
```

**Review by sayakpaul — APPROVED** [review]
This looks so NEAT!

@pacman100 for awareness.

File: docs/source/en/tutorials/using_peft_for_inference.md (hunk 2 above)
Comment (stevhliu) [reply docs/source/en/tutorials/using_peft_for_inference.md:168]: Oops sorry I forgot about this section! Added it back in as a nested section under the `fuse_lora` section 😅

File: docs/source/en/using-diffusers/merge_loras.md [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+pipeline = DiffusionPipeline.from_pretrained(
+    "username/fused-ikea-feng", torch_dtype=torch.float16,
+).to("cuda")
+
+pipeline.unet.to(memory_format=torch.channels_last)
+pipeline.unet = torch.compile(pipe.unet, mode="reduce-overhead", fullgraph=True)
+
+image = pipeline("A bowl of ramen shaped like a cute kawaii bear, by Feng Zikai", generator=torch.manual_seed(0)).images[0]
+image
+```
```
Comment (sayakpaul) [inline docs/source/en/using-diffusers/merge_loras.md:249]: I think this section is quite incomplete. 

1. It doesn't load any LoRA.
2. It doesn't apply fusion. 
3. It doesn't display the generated image. 

Could we please ensure the section is ported as closely as possible? `torch.compile()` compatibility with PEFT is super important for our users.

File: docs/source/en/tutorials/using_peft_for_inference.md (hunk 2 above)
Comment (sayakpaul) [reply docs/source/en/tutorials/using_peft_for_inference.md:168]: I am a little uncomfortable about how it's ported in the new section. 

Left a comment here: https://github.com/huggingface/diffusers/pull/7213/files#r1513791539.

**Review by sayakpaul — COMMENTED** [review]
Let's fix the `torch.compile()` section and then we can ship!

File: docs/source/en/using-diffusers/merge_loras.md (hunk 8 above)
Comment (stevhliu) [reply docs/source/en/using-diffusers/merge_loras.md:249]: About points 1-2:

Since the `torch.compile` section is already nested under the `fuse_lora` section, it is assumed that this is a subtopic of `fuse_lora` meaning you should read that section first. Adding the same code from the section above feels redundant. Do you think it'd be more helpful/clear if I added something saying "Make sure you read the`fuse_lora` section above first".

About point 3:

I think it'd be more impactful to show the speed up in time rather than another image. wdyt?

Comment (sayakpaul) [reply docs/source/en/using-diffusers/merge_loras.md]: I feel relatively strongly about points 1 - 2. Even if a single step is missed `torch.compile()` can fail during inference. 

In the `fuse_lora()` section, the step of the code shows `pipeline.unfuse_lora()`. This will make `torch.compile()` fail. The right sequence of the steps would be:

```python
# provided `set_adapters()` was already called. 
pipe.fuse_lora()
pipe.unload_lora_weights()

pipe.unet.to(memory_format=torch.channels_last)
pipe.unet = torch.compile(pipe.unet, mode="reduce-overhead", fullgraph=True)
``` 

So, I would appreciate it if the example code snippet for `torch.compile()` when doing LoRA fusion was as self-contained and as complete as possible. 

For pt. 3, the timings report will be inappropriate since it needs warming up. So, without that it will give an improper view of the timing.

Comment (stevhliu) [reply docs/source/en/using-diffusers/merge_loras.md]: Sounds good! I went with your suggestion to include an entirely self-contained code snippet 🙂

Comment (sayakpaul) [reply docs/source/en/using-diffusers/merge_loras.md]: Thank you Steven! Appreciate it.

**Comment by sayakpaul** [comment]
Feel free to merge it. Looking solid and thanks a mile for beating it with me.
